CI Fails on Windows Due to Platform-Specific Code and File Handling

@claude 

Issue: CI Fails on Windows Due to Platform-Specific Code and File Handling
Summary
The continuous integration (CI) workflow is failing on Windows runners due to a combination of platform-incompatible code (e.g., usage of os.statvfs, path separator logic, and platform-specific test assertions), file locking issues, and missing imports. See the full run and logs for details: [CI job run](https://github.com/quanticsoul4772/zeek-yara-integration/actions/runs/17339406907/job/49231179690) (ref: b61b277ddf77b7a9318f78eea8030d7baaeb40ff).

Failure Details
1. Platform-Specific Test Failures
Tests meant for Linux/macOS are failing on Windows (e.g., assert 'Windows' == 'Linux', path separator checks).
Usage of os.statvfs throws AttributeError on Windows.
Path separator checks fail ('\\' == '/').
2. File Handling and Directory Errors
NotADirectoryError: [WinError 267] when trying to access log files or directories that may not exist.
PermissionError: [WinError 32] caused by files being accessed by multiple processes or not being closed before deletion.
3. Missing Imports
NameError: name 'Console' is not defined in test_basic.py.
4. Faulty and Flaky Test Assertions
Error assertions such as assert None is not None and strict timestamp comparisons (e.g., assert x > x).
Tests expect error fields to always be populated; this may not be robust across platforms or Python versions.
Example Error Output
Code
FAILED tests/unit_tests/test_basic.py::test_imports - NameError: name 'Console' is not defined
FAILED tests/unit_tests/test_cleanup_manager.py::TestFileCleanupManager::test_check_disk_usage_alerts - AttributeError: <module 'os'> does not have the attribute 'statvfs'
FAILED tests/unit_tests/test_logging_utils.py::test_setup_logging_file_handler - NotADirectoryError: [WinError 267] The directory name is invalid: 'C:\\Users\\...\\exception.log'
FAILED tests/unit_tests/test_cleanup_manager.py::TestFileCleanupManager::test_cleanup_processed_files_success - PermissionError: [WinError 32] The process cannot access the file because it is being used by another process
...
FAILED tests/unit_tests/test_platform_markers.py::TestPlatformMarkers::test_linux_specific_behavior - AssertionError: This test should only run on Linux
...
Proposed Solutions
Platform-Specific Code & Tests
Use pytest.mark.skipif to skip platform-specific tests on incompatible systems.
Python
import sys
import pytest

@pytest.mark.skipif(sys.platform != "linux", reason="Linux only")
def test_linux_specific_behavior(): ...
Replace os.statvfs with shutil.disk_usage, which is cross-platform:
Python
import shutil

def get_disk_usage(directory):
    total, used, free = shutil.disk_usage(directory)
    return {'total': total, 'used': used, 'free': free}
Use os.path.sep for path separator logic, and avoid hardcoded checks.
File and Directory Handling
Ensure all directories exist before creating/opening files:
Python
import os

os.makedirs(os.path.dirname(log_file), exist_ok=True)
Make sure all file handles are closed before deletion or further access.
Use context managers (with open(...) as f:) to manage file handles.
Missing Imports
Add any missing imports, e.g., from rich.console import Console in test_basic.py.
Test Logic Improvements
Avoid strict or tautological assertions, e.g., replace assert x > x with more meaningful checks.
Ensure error assertions handle None gracefully, or check for expected error messages/codes.
References
[CI job logs, workflow, and errors (ref: b61b277ddf77b7a9318f78eea8030d7baaeb40ff)](https://github.com/quanticsoul4772/zeek-yara-integration/actions/runs/17339406907/job/49231179690)
Action Items
Refactor platform-specific tests to use conditional skips.
Replace or guard platform-unsupported functions.
Fix file opening/creation logic and ensure proper closing before deletion.
Add missing imports and review test assertions for robustness.

--- a/PLATFORM/core/cleanup_manager.py
+++ b/PLATFORM/core/cleanup_manager.py
@@ -11,6 +11,7 @@ configurable retention policies and processing status.
 import datetime
 import logging
 import os
+import shutil
 import threading
 import time
 from pathlib import Path
@@ -377,11 +378,8 @@ class FileCleanupManager:
             if not self.extract_dir.exists():
                 return None
 
-            stat = os.statvfs(str(self.extract_dir))
-
-            total_bytes = stat.f_frsize * stat.f_blocks
-            available_bytes = stat.f_frsize * stat.f_available
-            used_bytes = total_bytes - available_bytes
+            # Use cross-platform shutil.disk_usage instead of os.statvfs
+            total_bytes, used_bytes, available_bytes = shutil.disk_usage(str(self.extract_dir))
 
             usage_percent = (used_bytes / total_bytes) * 100 if total_bytes > 0 else 0
 

--- a/PLATFORM/utils/logging_utils.py
+++ b/PLATFORM/utils/logging_utils.py
@@ -101,7 +101,7 @@ def setup_logging(config, logger_name="zeek_yara"):
 
         # Ensure log directory exists
         log_dir = os.path.dirname(log_file)
-        if log_dir:  # Only create if dirname is not empty
+        if log_dir and os.path.isabs(log_file) and log_dir != log_file:  # Validate path is non-empty, absolute, and not the same as file
             os.makedirs(log_dir, exist_ok=True)
 
         # Create handler with rotation

--- a/PLATFORM/utils/logging_utils.py
+++ b/PLATFORM/utils/logging_utils.py
@@ -100,7 +100,9 @@ def setup_logging(config, logger_name="zeek_yara"):
         log_file = config["LOG_FILE"]
 
         # Ensure log directory exists
-        os.makedirs(os.path.dirname(log_file), exist_ok=True)
+        log_dir = os.path.dirname(log_file)
+        if log_dir:  # Only create if dirname is not empty
+            os.makedirs(log_dir, exist_ok=True)
 
         # Create handler with rotation
         file_handler = logging.handlers.RotatingFileHandler(

--- a/src/help_system.py
+++ b/src/help_system.py
@@ -21,6 +21,10 @@ try:
     RICH_AVAILABLE = True
 except ImportError:
     RICH_AVAILABLE = False
+    
+    # Define dummy Console class for type hints when rich is not available
+    class Console:
+        pass
 
 
 @dataclass

--- a/src/setup_wizard.py
+++ b/src/setup_wizard.py
@@ -25,6 +25,10 @@ try:
     RICH_AVAILABLE = True
 except ImportError:
     RICH_AVAILABLE = False
+    
+    # Define dummy Console class for type hints when rich is not available
+    class Console:
+        pass
 
 
 class SystemDetector:

--- a/src/tutorial_system.py
+++ b/src/tutorial_system.py
@@ -20,6 +20,10 @@ try:
     RICH_AVAILABLE = True
 except ImportError:
     RICH_AVAILABLE = False
+    
+    # Define dummy Console class for type hints when rich is not available
+    class Console:
+        pass
 
 
 @dataclass

--- a/tests/conftest.py
+++ b/tests/conftest.py
@@ -24,6 +24,32 @@ def pytest_configure(config):
     )
 
 
+def pytest_collection_modifyitems(config, items):
+    """Skip platform-specific tests on incompatible platforms."""
+    current_platform = platform.system()
+    
+    for item in items:
+        # Skip Linux-specific tests on non-Linux platforms
+        if item.get_closest_marker("linux") and current_platform != "Linux":
+            item.add_marker(pytest.mark.skip(reason="Linux only test"))
+        
+        # Skip macOS-specific tests on non-macOS platforms
+        if item.get_closest_marker("macos") and current_platform != "Darwin":
+            item.add_marker(pytest.mark.skip(reason="macOS only test"))
+        
+        # Skip Windows-specific tests on non-Windows platforms
+        if item.get_closest_marker("windows") and current_platform != "Windows":
+            item.add_marker(pytest.mark.skip(reason="Windows only test"))
+        
+        # Skip Unix-specific tests on non-Unix platforms
+        if item.get_closest_marker("unix") and current_platform not in ["Linux", "Darwin"]:
+            item.add_marker(pytest.mark.skip(reason="Unix-like systems only test"))
+        
+        # Skip POSIX-specific tests on non-POSIX platforms
+        if item.get_closest_marker("posix") and current_platform not in ["Linux", "Darwin"]:
+            item.add_marker(pytest.mark.skip(reason="POSIX-compliant systems only test"))
+
+
 @pytest.fixture
 def config():
     """Fixture for test configuration"""

--- a/tests/unit_tests/test_cleanup_manager.py
+++ b/tests/unit_tests/test_cleanup_manager.py
@@ -307,15 +307,11 @@ class TestFileCleanupManager(unittest.TestCase):
         self.assertTrue(status["matched"])
 
     @pytest.mark.unit
-    @patch("PLATFORM.core.cleanup_manager.os.statvfs")
-    def test_get_disk_usage(self, mock_statvfs):
+    @patch("PLATFORM.core.cleanup_manager.shutil.disk_usage")
+    def test_get_disk_usage(self, mock_disk_usage):
         """Test getting disk usage statistics"""
-        # Mock statvfs return value
-        mock_stat = Mock()
-        mock_stat.f_frsize = 4096  # 4KB blocks
-        mock_stat.f_blocks = 1000  # 1000 blocks total
-        mock_stat.f_available = 200  # 200 blocks available
-        mock_statvfs.return_value = mock_stat
+        # Mock shutil.disk_usage return value (total, used, free)
+        mock_disk_usage.return_value = (4096 * 1000, 4096 * 800, 4096 * 200)
 
         usage = self.cleanup_manager._get_disk_usage()
 
@@ -334,15 +330,14 @@ class TestFileCleanupManager(unittest.TestCase):
         self.assertIsNone(usage)
 
     @pytest.mark.unit
-    @patch("PLATFORM.core.cleanup_manager.os.statvfs")
-    def test_check_disk_usage_alerts(self, mock_statvfs):
+    @patch("PLATFORM.core.cleanup_manager.shutil.disk_usage")
+    def test_check_disk_usage_alerts(self, mock_disk_usage):
         """Test disk usage alerting"""
-        # Mock critical usage
-        mock_stat = Mock()
-        mock_stat.f_frsize = 4096
-        mock_stat.f_blocks = 1000
-        mock_stat.f_available = 50  # 95% usage
-        mock_statvfs.return_value = mock_stat
+        # Mock critical usage (95% used)
+        total_bytes = 4096 * 1000
+        used_bytes = 4096 * 950  # 95% usage
+        available_bytes = 4096 * 50
+        mock_disk_usage.return_value = (total_bytes, used_bytes, available_bytes)
 
         with patch.object(
             self.cleanup_manager.logger, "critical"

